dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/boogie/VisitorBoogie.java

```diff
@@ -739,57 +751,62 @@ public Object visitFun_expr(Fun_exprContext ctx) {
 
 	@Override
 	public Object visitIf_then_else_expr(If_then_else_exprContext ctx) {
-		BExpr guard = (BExpr)ctx.expr(0).accept(this);
-		IExpr tbranch = (IExpr)ctx.expr(1).accept(this);
-		IExpr fbranch = (IExpr)ctx.expr(2).accept(this);
-		return new IfExpr(guard, tbranch, fbranch);
+		Expression guard = (Expression) ctx.expr(0).accept(this);
+		Expression tbranch = (Expression) ctx.expr(1).accept(this);
+		Expression fbranch = (Expression) ctx.expr(2).accept(this);
+		return expressions.makeConditional(guard, tbranch, fbranch);
 	}
 
 	@Override
 	public Object visitParen_expr(Paren_exprContext ctx) {
 		return ctx.expr().accept(this);
 	}
 
-	@Override 
+	@Override
 	public Object visitBv_expr(Bv_exprContext ctx) {
 		String value = ctx.getText().split("bv")[0];
-		int precision = Integer.parseInt(ctx.getText().split("bv")[1]);
-		return new IValue(new BigInteger(value),precision);
+		int bitWidth = Integer.parseInt(ctx.getText().split("bv")[1]);
+		return expressions.parseValue(value, types.getIntegerType(bitWidth));
 	}
 
 	@Override
 	public Object visitInt_expr(Int_exprContext ctx) {
-		return new IValue(new BigInteger(ctx.getText()), getArchPrecision());
+		return expressions.parseValue(ctx.getText(), types.getArchType());
```

Comment:
Afaik boogie literals are `false`, `true`, `0`, `0bv32`, etc..  Smack defines `type i64 = int;` and `type ref = i64;` and inserts integer literals, i.e.
```boogie
$p2 := $add.ref($add.ref($p1, $mul.ref(0, 16)), $mul.ref(8, 1));
```
We could try to interpret `i32` as `bv32` in a future PR.

Reply:
Even if it ends up defining `iX = int` where `X` is the precision, can't we make use of this information ? As far as I understand, we know have precision event if we eventually end up encoding things with mathematical integers. Why not to make use of this information during parsing (instead of falling back to getArchPrecision?

Reply:
I agree that we should parse the `iX` as `bvX`. If it is not too much to change, we can do it in this PR. Otherwise I'm fine with putting that feature into a separate PR (+ some explotation of the new information).

Reply:
If we don't want the encoder to last-minute-fix type mismatches, the Boogie parser should account for them by implicitly casting integer expressions of all kinds.  In a uniform way, this entails that the following
```boogie
$mul.ref(24,$zext.i32.i64(1))
```
is parsed to `(mul (cast ref 24int) (cast ref (cast bv64 (cast bv32 1int) ) ) )`.

This means that there are probably some additional simplification rules to add, as well.  I will try this.